vortex-scan/src/lib.rs

```diff
@@ -330,5 +316,168 @@ fn to_field_mask(field: FieldName) -> FieldMask {
     FieldMask::Prefix(FieldPath::from(Field::Name(field)))
 }
 
+/// A projected subset (by indices, range, and filter) of rows from a Vortex data source.
+///
+/// The method of this struct enable, possibly concurrent, scanning of multiple row ranges of this
+/// data source.
+///
+/// See also: [ScanBuilder].
+pub struct RepeatedScan<A: 'static + Send> {
+    layout_reader: LayoutReaderRef,
+    projection: ExprRef,
+    filter: Option<ExprRef>,
+    /// Optionally read a subset of the rows in the file.
+    row_range: Option<Range<u64>>,
+    /// The selection mask to apply to the selected row range.
+    // TODO(joe): replace this is usage of row_id selection, see
+    selection: Selection,
+    /// The natural splits of the file.
+    splits: Vec<Range<u64>>,
+    /// The number of splits to make progress on concurrently **per-thread**.
+    concurrency: usize,
+    /// Function to apply to each [`ArrayRef`] within the spawned split tasks.
+    map_fn: Arc<dyn Fn(ArrayRef) -> VortexResult<A> + Send + Sync>,
+    /// Maximal number of rows to read (after filtering)
+    limit: Option<usize>,
+    /// The fields we will read.
+    #[allow(dead_code)]
+    field_mask: Vec<FieldMask>,
+    /// The dtype of the projected arrays.
+    dtype: DType,
+}
+
+pub fn intersect_ranges(
+    left: Option<&Range<u64>>,
+    right: Option<Range<u64>>,
+) -> Option<Range<u64>> {
+    match (left, right) {
+        (None, None) => None,
+        (None, Some(r)) => Some(r),
+        (Some(l), None) => Some(l.clone()),
+        (Some(l), Some(r)) => Some(cmp::max(l.start, r.start)..cmp::min(l.end, r.end)),
+    }
+}
+
+impl<A: 'static + Send> RepeatedScan<A> {
+    pub fn build(
+        &self,
+        row_range: Option<Range<u64>>,
+    ) -> VortexResult<Vec<BoxFuture<'static, VortexResult<Option<A>>>>> {
+        let row_range = intersect_ranges(self.row_range.as_ref(), row_range);
+
+        let ctx = Arc::new(TaskContext {
+            row_range,
+            selection: self.selection.clone(),
+            filter: self.filter.clone().map(|f| Arc::new(FilterExpr::new(f))),
+            reader: self.layout_reader.clone(),
+            projection: self.projection.clone(),
+            mapper: self.map_fn.clone(),
+        });
+
+        // Create a task that executes the full scan pipeline for each split.
+        let mut limit = self.limit;
+        let split_tasks = self
+            .splits
+            .iter()
+            .filter_map(|split_range| {
+                if limit.is_some_and(|l| l == 0) {
+                    None
+                } else {
+                    Some(split_exec(ctx.clone(), split_range.clone(), limit.as_mut()))
+                }
+            })
+            .try_collect()?;
+
+        Ok(split_tasks)
+    }
+
+    pub async fn build_single_threaded(
```

Comment:
This avoids futures completely, right? Maybe call it blocking? But does this actually have an impact comparing to setting concurrency 1?